### ISSUE TITLE
More Flexible Transport API for Updater

### ISSUE THREAD

**Issue by awwad** [theupdateframework/tuf#865]
### Purpose

Work here will have to wait until after #846 (changes to updater) and will motivate or follow work on #841 (updater refactor to expose better API and separate transport and verification).

This conversation is one we should have as we proceed with those changes.

The usage of HTTP(S) in the TUF reference implementation is for convenience (basic use cases, currently used in production IIUC) and demonstration.  Despite that, HTTP(S) is actually pretty deeply built into the reference implementation, which is unfortunate.

I think it's important that TUF design is _vociferously ambivalent to transport mechanism_.  Designs that use it -- like Uptane -- may support various and odd transport mechanisms.  Because of this designed ambivalence and diversity of use, it is beneficial even purely for reference purposes (not to mention production purposes) to have a clean separation between the [verification of metadata and targets] and the [acquisition of metadata and targets].  The current stack in `updater` does not do this well, as covered a bit in #841.

The good news is that this won't be too hard, and we have both breaking changes and other API updates coming soon, so this season is a good time. :)

### Initial Task / Plan
Beyond #841, I want an API in `updater` that exposes functions that:
- take targetinfo (as obtained by e.g. `get_one_valid_targetinfo`) and either bytes or a file object, obtained through whatever external mechanism, and verifies the bytes (hashes, length, etc.).
   - Note: This function must emphatically advise that data that is not verified should be destroyed, and that data that is not yet verified in this way be stored separately from verified data!
- obtain a temp file (via the existing http(s) transport mechanisms used) that has not been verified yet and returns the temp file object.  (or bytes?).
   - Slow transport attacks and infinite data attacks should be mitigated as they currently are (caveat for current slow transport attacks)
   - Note: This function must emphatically advise that data retrieved this way has NOT yet been verified.
- recover files via the sample http(s) transport mechanisms into temp files and verify them, discarding them if they are not verified  (current functionality, like `_get_metadata_file`, but exposed as discussed in #841)

In addition, work should be done to reorganize `updater` and present it as both a simple client tool, and a library for a variety of functions (like the above) that are of use to an updater implementation.  Notes should be made in the documentation.  Attention should be drawn in both docstrings and docs/tutorials to the fact that separate, additional custom verification can be performed, but that it should not be at the expense of the existing verification process.  There may come a time when we need to separate the sample-client elements (like the 3rd example API function above) from the core API elements (like the first two API functions above)

**Comment by renatav** [comment]
I think that this, when implemented, will be of big use to us.

We had to modify `_get_metadata_file` and replace `tuf.download.unsafe_download(file_mirror, upperbound_filelength)` with our own implementation of the download step. [Here's the PR](https://github.com/openlawlibrary/tuf/pull/3/files) in our local fork where we made it possible to redefine download of metadata files. Targets are handled similarly to metadata files, but we had to redefine the file digest calculation step, so that the file is read from a git repository.

Another problem that we faced while implementing validation of git repositories is that we do not simply want to validate the top commit. We want to check all commits that were pushed to remote since the client updated their repository. That means that we need to repeat the complete update process (refresh metadata, download and validate targets) several times with different mirrors. This is something that we can find a workaround for though.

Some other problems that we faced, but are not directly related to this particular issue are the following:
- Because we want to validate older commits, we had to redefine validation of the expiration date. We want to check if a file was valid when it was committed.
- We needed a check which makes sure that if a file should remain unchanged according to the reference metadata, it actually remained the same.
